src/puppeteer_pool.js

```diff
@@ -388,44 +399,70 @@ class PuppeteerPool {
         });
     }
 
+    /**
+     * Updates the instance metadata when a new page is opened.
+     *
+     * @param {PuppeteerInstance} instance
+     * @ignore
+     */
+    _updateInstance(instance) {
+        instance.lastPageOpenedAt = Date.now();
+        instance.totalPages++;
+        if (instance.totalPages >= this.retireInstanceAfterRequestCount) this._retireInstance(instance);
+    }
+
+    /**
+     * Produces a new page instance either by reusing an idle page that currently isn't processing
+     * any request or by spawning a new page (new browser tab) in one of the available
+     * browsers when no idle pages are available.
+     *
+     * To spawn a new browser tab for each page, set the `reusePages` constructor option to false.
+     *
+     * @return {Promise<Page>}
+     */
+    async newPage() {
+        let idlePage;
+        // We don't need to check whether options.reusePages is true,
+        // because if it's false, the array will be empty and the loop will never start.
+        while (idlePage = this.idlePages.shift()) { // eslint-disable-line no-cond-assign
+            // Since pages can close for various reasons that we have no control over,
+            // we need to make sure that we're only using live pages, so we go
+            // through the queue until we get a page that's live, which means
+            // that it's not closed and its browser is not retired.
+            const pageIsNotClosed = !this.closedPages.has(idlePage);
```

Comment:
I would consider to remove page from `idlePages` at the moment when either browser is marked retired or page is added to `closedPages`. I am not sure how complicated this would be but it's the place where this logic belongs. 

Then this function would be just `return this.idlePages.shift() || this._openNewTab()`

Reply:
Yeah, I considered this, but you can't really delete using the object reference, so you'd first have to find the page in the array and then delete it using an index, which would create a sparse array which are super unoptimized in V8, so I just went with the current implementation. 

Or we could create a custom data structure that does that right, but that seems like an overkill...